[ONNX] Public diagnostic options for 'dynamo_export'

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107165
* #107158
* __->__ #106741

Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

BowenBao committed:
[ONNX] Public diagnostic options for 'dynamo_export'

[ghstack-poisoned]

BowenBao committed:
Update according to discussions on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

1. Choose if you want to produce a .sarif file and specify its location.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.


[ghstack-poisoned]

BowenBao commented:
Unfortunately tagged a lot of people due to changes in `benchmarks/dynamo/common.py` due to an onnx exporter api update.

BowenBao committed:
Update on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

BowenBao committed:
rebase w/ dispatcher change on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

BowenBao committed:
remove old code on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

BowenBao committed:
Fix CI, rebase, address comments on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -879,8 +936,7 @@ def _assert_fake_tensor_mode(self):
         ) and not self.options.fake_context:
-            return OnnxExporterError(
-                self.options.diagnostic_context,
+            return RuntimeError(
```
There was a bug using "return" instead of "raise". I cannot correct it as `test_fx_symbolic_tracer_large_scale_exporter` tests start to fail if I do. @thiagocrepaldi please take a look later.

thiagocrepaldi replied:
Please create an issue and assigne me or maybe @titaiwangms who will help with some fake tasks

BowenBao committed:
lint, doc on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -76,6 +78,11 @@
 _DEFAULT_FAILED_EXPORT_SARIF_LOG_PATH = "report_dynamo_export.sarif"
 """The default path to write the SARIF log to if the export fails."""
 
+log = logging.getLogger(__name__)
```
Why moved out from `ExportOptions`?

Doesn't this invalidate the premises that `Exporter` owns everything that the export needs and it is thread safe (although we don't care about thread safety lol)

One consequence of this is that `import torch` now pays the compute time to execute `loggin.getLogger` which is not strictly necessary during import time

BowenBao replied:
This is the more idiomatic way of how logger is used. Our way of owning logger instance is actually listed as [pattern to avoid](https://docs.python.org/3/howto/logging-cookbook.html#using-loggers-as-attributes-in-a-class-or-passing-them-as-parameters). 

Meanwhile, the more important reason is we are setting static logger name as the one registered with PT2 logging infra. It no longer makes sense to allow users creating and passing in their own logger.

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -772,6 +781,54 @@ def save(
+        self.diagnostic_context.dump(destination)
+
+    @classmethod
+    def from_export_failure(
```
Does it have to be public? If so, add a unit test that test it

I only saw `dynamo_export` use it internally, so it could be private

thiagocrepaldi replied:
```suggestion
    def from_failure(
```
This is a method of `ExportOutput`, which implies `export` already.

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -772,6 +781,54 @@ def save(
                 serializer.serialize(self, destination)
 
+    @_beartype.beartype
+    def save_sarif_log(self, destination: str) -> None:
```
```suggestion
    def save_sarif(self, destination: str) -> None:
```
Rename for discussion, only.
Is it implied that SARIF is always a log file? If so, we can save the `_log` suffix. Or is there `sarif_another_type_of_artifact` possible in the sarif world?

BowenBao replied:
I think your suggestion is more concise.

thiagocrepaldi replied:
i like `save_sarif` better because whoever doesnt know what SARIF is, has no reason to `save_sarif`
However, who does, may not know they can get the sarif from `save_diagnostics`.

BowenBao replied:
`save_diagnostics_to_sarif` then?

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -1023,14 +1083,18 @@ def dynamo_export(
-        # the issue to be added.
         message = (
             f"Failed to export the model to ONNX. Generating SARIF report at {sarif_report_path}. "
+            f"SARIF is a standard format for the output of static analysis tools. "
+            f"SARIF log can be loaded in VS Code SARIF viewer extension, "
+            f"or SARIF web viewer(https://microsoft.github.io/sarif-web-component/)."
```
```suggestion
            "SARIF is a standard format for the output of static analysis tools. "
            "SARIF log can be loaded in VS Code SARIF viewer extension, "
            "or SARIF web viewer(https://microsoft.github.io/sarif-web-component/)."
```
no variable expansion is needed for these lines, so we can save the compute

BowenBao replied:
surprised not flagged by linter

BowenBao committed:
address comments on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

titaiwangms commented on test/onnx/dynamo/test_exporter_api.py:
```diff
@@ -137,6 +124,43 @@ def forward(self, x):
+        self.assertIsInstance(cm.exception, torch.onnx.OnnxExporterError)
+        self.assertIsInstance(cm.exception.export_output, ExportOutput)
+
+    def test_access_export_output_model_proto_raises_when_export_output_is_emitted_from_failed_export(
```
```suggestion
    def test_access_model_proto_raises_runtime_error_when_export_output_is_emitted_from_failed_export
```

BowenBao commented on test/onnx/dynamo/test_exporter_api.py:
```diff
@@ -137,6 +124,43 @@ def forward(self, x):
+            dynamo_export(ModelWithExportError(), torch.randn(1, 1, 2))
+        export_output = cm.exception.export_output
+        with self.assertRaises(RuntimeError):
+            export_output.model_proto
```
Wanted to keep signature clean without `Optional`.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -772,6 +781,56 @@ def save(
+        self.diagnostic_context.dump(destination)
+
+    @classmethod
+    def _from_failure(
```
It is internal details of ExportOutput on how to instantiate some fields (e.g., model_proto) when failure.

BowenBao committed:
Update on "[ONNX] Public diagnostic options for 'dynamo_export'"


Generate diagnostic reports to monitor the internal stages of the export process. This tool aids in unblocking model exports and debugging the exporter.

#### Settings

~~1. Choose if you want to produce a .sarif file and specify its location.~~
1. Updated: saving .sarif file should be done by `export_output.save_sarif_log(dst)`, similar to saving exported onnx model `export_output.save(model_dst)`.
2. Customize diagnostic options:
    - Set the desired verbosity for diagnostics.
    - Treat warnings as errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -898,17 +956,10 @@ def _assert_fake_tensor_mode(self):
         ) and self.options.fake_context:
-            raise OnnxExporterError(
-                self.options.diagnostic_context,
+            raise RuntimeError(
```
I didn't follow why the error is changed? Just for my education. Thanks!

BowenBao replied:
We are now making `OnnxExporterError` the top level user facing error, encapsulating an `ExportOutput` instance. 

Hence, any error raised from within exporter should raise in its own corresponding error type. Top level export call will catch these and encapsulate accordingly.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -772,6 +781,56 @@ def save(
                 serializer.serialize(self, destination)
 
+    @_beartype.beartype
+    def save_diagnostics(self, destination: str) -> None:
```
Overall I don't have strong opinion.

BowenBao commented:
@malfet could you help approve this one? Had a 1 line change in `benchmarks/dynamo/common.py` due to onnx export api changes.
